File: week-4/sudoku-solver/sudoku-solver.py
class Solution:
    def solveSudoku(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        self.solveBoard(0, 0, board)
    
    def solveBoard(self, row, col, board):
        if row == len(board):
            return True
        if col == len(board):
            return self.solveBoard(row+1, 0, board)
        if board[row][col] != '.':
            return self.solveBoard(row, col+1, board)
        
        for i in range(1, 10):
            value = str(i)
            if self.validate(row, col, value, board):
                board[row][col] = value
                if self.solveBoard(row, col+1, board):
                    return True
                board[row][col] = '.'
        
        return False
    
    def validate(self, row, col, number, board):
        subRow = (row//3) * 3
        subCol = (col//3) * 3
        
        for i in range(len(board)):
            if board[row][i] == number or board[i][col] == number:
                return False
            
        for i in range(subRow, subRow+3):
            for j in range(subCol, subCol+3):
                if board[i][j] == number:
                    return False
                
        return True

File: week-4/sudoku-solver/sudoku-solver.js
/**
 * @param {character[][]} board
 * @return {void} Do not return anything, modify board in-place instead.
 */
var solveSudoku = function(board) {
    solveBoard(0, 0, board);
};

const solveBoard = (row, col, board) => {
    if (row === board.length) return true;
    if (col === board.length) return solveBoard(row+1, 0, board);
    if (board[row][col] != '.') return solveBoard(row, col+1, board);
    
    for (let i=1; i<10; i++) {
        const value = String(i);
        if (validate(row, col, value, board)) {
            board[row][col] = value;
            if (solveBoard(row, col+1, board)) return true;
            board[row][col] = '.';
        }
    }
        
    return false;
}

const validate = (row, col, number, board) => {
    const subRow = Math.floor(row/3)*3;
    const subCol = Math.floor(col/3)*3;
    
    for (let i=0; i<board.length; i++) {
        if (board[row][i] === number || board[i][col] === number) return false;
    }
    
    for (let i=subRow; i<subRow+3; i++) {
        for (let j=subCol; j<subCol+3; j++) {
            if (board[i][j] === number) return false;
        }
    }
    return true;
}
